quiche/src/path.rs

```diff
@@ -57,15 +57,83 @@ pub enum PathState {
     Validated,
 }
 
-impl PathState {
+impl PathValidationState {
     #[cfg(feature = "ffi")]
     pub fn to_c(self) -> libc::ssize_t {
         match self {
-            PathState::Failed => -1,
-            PathState::Unknown => 0,
-            PathState::Validating => 1,
-            PathState::ValidatingMTU => 2,
-            PathState::Validated => 3,
+            PathValidationState::Failed => -1,
+            PathValidationState::Unknown => 0,
+            PathValidationState::Validating => 1,
+            PathValidationState::ValidatingMTU => 2,
+            PathValidationState::Validated => 3,
+        }
+    }
+}
+
+/// The different usage states of the path.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+enum PathState {
+    /// The path only sends probing packets.
+    Unused,
+    /// The path can send non-probing packets.
+    Active,
+    /// The path is under closing process.
+    Closing(u64, Vec<u8>),
+    /// The path is now closed.
+    Closed(u64, Vec<u8>),
+}
+
+/// The different requests that can be assigned to a path.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PathRequest {
+    /// The path should not send non-probing packets.
+    Unused,
+    /// The path should send probing packets.
+    Active,
+    /// The path should be abandonned, with the provided error code and reason
+    /// message.
+    Abandon(u64, Vec<u8>),
+}
+
+impl PathRequest {
+    fn requested_state(self) -> PathState {
+        match self {
+            PathRequest::Unused => PathState::Unused,
+            PathRequest::Active => PathState::Active,
+            PathRequest::Abandon(e, r) => PathState::Closing(e, r),
+        }
+    }
+}
+
+/// The status of a path, advertised through the PATH_STATUS frame.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum PathStatus {
+    /// The host should stop sending non-probing packets on the path.
+    Standby,
```

Comment:
Should `PathStatus` we receive from the peer influence usability of paths? AFAICT the `usable()` method only takes into account `PathState`, which isn't updated when `PathStatus` is received.

Reply:
Taking into account the `PathStatus` into account is somewhat a scheduling decision that should be driven by the application. The default `get_send_path_id()` (called when using `send()` instead of `send_on_path()`) takes the `PathStatus` into account (it prefers all `Available` paths when possible, and only consider `Standby` ones when there only remains `Standby` paths).

A validated `Standby` path is by itself usable, there is some policy stating that "if it is possible, do not use that path".  Directly taking the `PathStatus` into account in the `usable()` method is probably not desirable, especially if all the remaining paths are `Standby` ones.